pkg/operator/controllers/machineset/machineset_controller.go

```diff
@@ -0,0 +1,95 @@
+package machineset
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"regexp"
+
+	"github.com/Azure/go-autorest/autorest/to"
+	machinev1beta1 "github.com/openshift/machine-api-operator/pkg/apis/machine/v1beta1"
+	maoclient "github.com/openshift/machine-api-operator/pkg/generated/clientset/versioned"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	aroclient "github.com/Azure/ARO-RP/pkg/operator/clientset/versioned"
+	"github.com/Azure/ARO-RP/pkg/operator/controllers"
+)
+
+type MachineSetReconciler struct {
+	log    *logrus.Entry
+	maocli maoclient.Interface
+	arocli aroclient.Interface
+}
+
+func NewMachineSetReconciler(log *logrus.Entry, maocli maoclient.Interface, arocli aroclient.Interface) *MachineSetReconciler {
+	return &MachineSetReconciler{
+		log:    log,
+		maocli: maocli,
+		arocli: arocli,
+	}
+}
+
+func (r *MachineSetReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	instance, err := r.arocli.AroV1alpha1().Clusters().Get(ctx, arov1alpha1.SingletonClusterName, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	machinesetObject, err := r.maocli.MachineV1beta1().MachineSets(machineSetsNamespace).Get(ctx, request.Name, metav1.GetOptions{}) // need to re-add workers only
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Count amount of current worker replicas
+	replicaCount := 0
+	machinesets, err := r.maocli.MachineV1beta1().MachineSets(machineSetsNamespace).List(ctx, metav1.ListOptions{LabelSelector: "machine.openshift.io/cluster-api-machine-role=worker"})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	for _, machineset := range machinesets.Items {
+		if machineset.Spec.Replicas != nil {
+			replicaCount += int(*machineset.Spec.Replicas)
+			r.log.Infof("Worker count is %v for InfraID '%v'", replicaCount, instance.Spec.InfraID)
+		}
+	}
+
+	// Verify that MachineSet InfraID matches cluster InfraID
+	matches, err := regexp.Match(instance.Spec.InfraID, []byte(machinesetObject.ObjectMeta.Name))
```

Comment:
Since we're counting all worker replicas, we need this or something like this as a guard rail against us modifying customer MachineSets, but it's definitely not perfect. For example, there could be a custom MachineSet with the same name and cluster label as ours, and we wouldn't know the difference. I'm open to suggestions (or feedback on whether or not we should care about this).